File: lua/taskui/events.lua
-- lua/taskui/events.lua
--
-- A socket the terminal's taskui reports to, and what Neovim does about it.
--
-- The terminal is the interface — the whole terminal UI, with its pivots, its
-- folds, its archive, none of it reimplemented here. What Neovim adds is the
-- part a terminal cannot do from inside itself: land on the failing line, keep
-- the failure in the statusline while you fix it, and open the file `e` chose
-- in the editor that is already open rather than in one nested inside it.

local config = require("taskui.config")

local M = {}

--- runs is what the events have said so far, by root task name. Nothing here is
--- rendered; it is what the statusline and the quickfix hook read.
---@type table<string, table>
M.runs = {}

--- newest is the run to speak for: whatever is going, or the last thing to
--- finish. The statusline shows one thing, and this is which.
M.newest = nil

local server, path

--- Handlers by event type. Anything unknown is ignored on purpose: a newer
--- binary saying more than this plugin understands must not be an error.
local handlers = {}

function handlers.run(event)
  M.runs[event.root] = {
    root = event.root,
    dir = event.dir,
    status = "running",
    started = event.started_unix,
    tasks = {},
    failed = {},
  }
  M.newest = M.runs[event.root]
end

function handlers.task(event)
  local run = M.runs[event.root]
  if not run then
    return
  end
  run.tasks[event.name] = { status = event.status, duration_ms = event.duration_ms }
  if event.status == "Failed" then
    run.failed[event.name] = true
  end
end

function handlers.exit(event)
  local run = M.runs[event.root]
  if not run then
    return
  end
  run.status = (event.code == 0) and "ok" or "failed"
  run.exit = event.code
  run.duration_ms = event.duration_ms
  run.saved = event.saved
  M.newest = run
  M.finished(run)
end

function handlers.edit(event)
  -- `e` in the terminal, opened in the editor hosting it. Without this the TUI
  -- would launch $EDITOR inside its own window, which is a second Neovim in a
  -- terminal inside the first one.
  require("taskui.term").open_file(event.path, event.line, event.col)
  if event.note and event.note ~= "" then
    vim.notify("taskui: " .. event.note, vim.log.levels.WARN)
  end
end

--- What happens when a run ends: the quickfix list, and a word about how it
--- went — because the terminal may not be the window you are looking at.
---@param run table
function M.finished(run)
  local failed = run.status == "failed"
  local when = config.options.quickfix
  if when == "always" or (when == "on_failure" and failed) then
    require("taskui.cli").quickfix(run.root, function(items, err)
      if err or #items == 0 then
        return
      end
      vim.fn.setqflist({}, " ", { title = "taskui · " .. run.root, items = items })
      if failed and config.options.open_quickfix then
        vim.cmd("copen")
      end
    end)
  end
  if config.options.notify then
    local took = M.duration(run.duration_ms)
    if failed then
      vim.notify(
        ("taskui: %s failed (exit %s)%s"):format(run.root, tostring(run.exit), took),
        vim.log.levels.ERROR
      )
    else
      vim.notify(("taskui: %s passed%s"):format(run.root, took))
    end
  end
end

--- A duration read at a glance rather than in full: ` in 4ms`, ` in 1.5s`.
---@param ms integer|nil
---@return string
function M.duration(ms)
  if not ms or ms <= 0 then
    return ""
  end
  if ms < 1000 then
    return (" in %dms"):format(ms)
  end
  local seconds = ms / 1000
  if seconds < 60 then
    return (" in %.1fs"):format(seconds)
  end
  return (" in %dm%02ds"):format(math.floor(seconds / 60), math.floor(seconds % 60))
end

--- Feeds one line of the stream. Exposed so the tests can drive it without a
--- socket, and because a malformed line must be skipped rather than thrown.
---@param line string
function M.feed(line)
  if line == "" then
    return
  end
  local ok, event = pcall(vim.json.decode, line)
  if not ok or type(event) ~= "table" or not event.type then
    return
  end
  local handler = handlers[event.type]
  if handler then
    handler(event)
  end
end

--- Starts listening, and returns the socket path to hand the terminal. One
--- server for the session: every taskui started from this Neovim reports to
--- the same place, and the events carry which run they are about.
---@return string|nil path
function M.listen()
  if server and path then
    return path
  end
  path = vim.fn.tempname() .. ".taskui.sock"
  server = vim.uv.new_pipe(false)
  local ok, err = server:bind(path)
  if not ok then
    vim.notify("taskui: could not listen for events: " .. tostring(err), vim.log.levels.WARN)
    server, path = nil, nil
    return nil
  end
  server:listen(8, function()
    local client = vim.uv.new_pipe(false)
    server:accept(client)
    -- The stream arrives in chunks that split wherever the pipe did, so the
    -- tail of a chunk is only sometimes a whole line.
    local rest = ""
    client:read_start(function(read_err, chunk)
      if read_err or not chunk then
        client:close()
        return
      end
      rest = rest .. chunk
      while true do
        local at = rest:find("\n", 1, true)
        if not at then
          return
        end
        local line = rest:sub(1, at - 1)
        rest = rest:sub(at + 1)
        vim.schedule(function()
          M.feed(line)
        end)
      end
    end)
  end)
  return path
end

--- Stops listening and removes the socket.
function M.stop()
  if server then
    server:close()
    server = nil
  end
  if path then
    os.remove(path)
    path = nil
  end
end

return M

File: internal/events/events.go
// Package events is taskui's run, addressed to another program.
//
// Two callers, one shape. `--run <task> --json` writes the stream to stdout and exits with
// the run; `--events <path>` writes it from inside the TUI to a socket somebody else is
// listening on, so an editor hosting the terminal can fill a quickfix list, colour a
// statusline and open the file a failure named — without reimplementing any of it.
//
// The events are differences against a run's own state rather than the engine's internal
// event stream. That is deliberate: [github.com/ddromanidis/taskui/internal/run.Run.Poll]
// folds raw events into state — a line arriving is also a task starting — and a consumer
// wants the folded version. "Task went from pending to running" has no event of its own and
// is exactly what a front end draws.
package events

import (
	"encoding/json"
	"errors"
	"io"
	"net"
	"os"
	"sync"

	"github.com/ddromanidis/taskui/internal/run"
)

// The events, one JSON object per line. Separate types rather than one struct with
// everything in it: a protocol whose every line carries `"secrets": 0` is a protocol nobody
// enjoys reading, and `omitempty` on a line index would drop the first line of every task.
type (
	// RunStarted opens a run's events. Everything after it that names this root belongs to
	// this run, until Exit.
	RunStarted struct {
		Type    string   `json:"type"` // "run"
		Root    string   `json:"root"`
		Dir     string   `json:"dir"`
		Args    []string `json:"args,omitempty"`
		Started int64    `json:"started_unix"`
	}
	// Graph is the execution graph, sent once, as soon as it is resolved.
	Graph struct {
		Type  string              `json:"type"` // "graph"
		Root  string              `json:"root"`
		Edges map[string][]string `json:"edges"`
	}
	// Task is a task's status changing, and its duration once that is known.
	Task struct {
		Type string `json:"type"` // "task"
		Root string `json:"root"`
		Name string `json:"name"`
		// Status is Pending, Running, Ok, Failed or Skipped — the names the archive uses.
		Status     string `json:"status"`
		DurationMs int64  `json:"duration_ms,omitempty"`
		Note       string `json:"note,omitempty"`
	}
	// Line is one line of output.
	Line struct {
		Type string `json:"type"` // "line"
		Root string `json:"root"`
		Task string `json:"task"`
		// Index counts from the start of the run, not from the start of the buffer: a task
		// that outlives the 20,000-line cap drops lines off the front, and an index that
		// slid with them would renumber everything a consumer had already stored.
		Index   int    `json:"index"`
		Text    string `json:"text"`
		Command bool   `json:"command,omitempty"`
	}
	// Prompt is a task waiting to be answered, which a host may want to say out loud: a
	// run that appears to hang is usually one asking a question.
	Prompt struct {
		Type string `json:"type"` // "prompt"
		Root string `json:"root"`
		Text string `json:"text"`
	}
	// Exit ends a run.
	Exit struct {
		Type       string `json:"type"` // "exit"
		Root       string `json:"root"`
		Code       int    `json:"code"`
		DurationMs int64  `json:"duration_ms"`
		Saved      string `json:"saved,omitempty"`
		Secrets    int    `json:"redacted_secrets,omitempty"`
	}
	// Edit is `e` pressed on a location. A host editor opens it; without one, taskui
	// launches $EDITOR itself.
	Edit struct {
		Type string `json:"type"` // "edit"
		Path string `json:"path"`
		Line int    `json:"line"`
		Col  int    `json:"col,omitempty"`
		// Note is why this is not quite where you asked for — an ambiguous name resolved
		// to a guess, a fallback to the task's first location.
		Note string `json:"note,omitempty"`
	}
)

// Sink is where events go. A nil Sink is a working Sink that drops everything, so a caller
// with no host attached needs no branch of its own.
type Sink struct {
	mu  sync.Mutex
	w   io.WriteCloser
	enc *json.Encoder
	// Lines says whether output lines go out too. The `--json` stream sends them, because
	// its consumer is drawing the run; a TUI with a host attached does not, because the
	// terminal in front of you is already showing them.
	Lines bool
}

// Open dials a destination for events.
//
// A unix socket first, which is what a host listening for them offers: a socket says when
// the other end went away, needs no cleanup, and cannot be read half-written. Failing that
// the path is opened as a file, appended to, which is what a shell redirection and a test
// both want.
func Open(path string) (*Sink, error) {
	if path == "" {
		return nil, errors.New("no path to report events to")
	}
	if conn, err := net.Dial("unix", path); err == nil {
		return New(conn), nil
	}
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return nil, err
	}
	return New(f), nil
}

// New wraps an already-open destination.
func New(w io.WriteCloser) *Sink {
	return &Sink{w: w, enc: json.NewEncoder(w)}
}

// Send writes one event. Errors are dropped on purpose: a host that closed the socket has
// stopped caring, and a run must not fail because nobody is listening any more.
func (s *Sink) Send(event any) {
	if s == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_ = s.enc.Encode(event)
}

// Close releases the destination.
func (s *Sink) Close() error {
	if s == nil {
		return nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.w.Close()
}

// Deltas remembers what has already gone out about one run, so each poll sends only what is
// new. One per run: two runs at once are two of these, which is what keeps a busy slot from
// renumbering a quiet one.
type Deltas struct {
	// seen is how many lines of each task have been sent, counted from the start of the
	// run so that lines dropped off the front of the buffer do not renumber anything.
	seen   map[string]int
	said   map[string]taskState
	graph  bool
	opened bool
	closed bool
	prompt string
}

// taskState is what was last said about a task. The duration is part of it because it
// arrives after the status does — a task is Ok some milliseconds before the clock settles —
// and a consumer told `Ok` with no duration would never hear the figure.
type taskState struct {
	status     run.Status
	durationMs int64
}

// NewDeltas starts tracking a run.
func NewDeltas() *Deltas {
	return &Deltas{seen: map[string]int{}, said: map[string]taskState{}}
}

// Start sends the run's opening event, once.
func (d *Deltas) Start(s *Sink, r *run.Run, dir string) {
	if d.opened {
		return
	}
	d.opened = true
	s.Send(RunStarted{
		Type: "run", Root: r.Root, Dir: dir,
		Args: r.Args, Started: r.Started.Unix(),
	})
}

// Flush sends everything that has changed since the last call.
func (d *Deltas) Flush(s *Sink, r *run.Run) {
	if s == nil || r == nil {
		return
	}
	if !d.graph && len(r.Graph.Edges) > 0 {
		d.graph = true
		s.Send(Graph{Type: "graph", Root: r.Root, Edges: r.Graph.Edges})
	}

	for _, name := range r.TaskNames() {
		task := r.Tasks[name]
		if task == nil || name == "" {
			continue
		}
		now := taskState{status: task.Status, durationMs: task.Duration().Milliseconds()}
		was, known := d.said[name]
		changed := !known || was != now
		announce := func() {
			d.said[name] = now
			s.Send(Task{
				Type: "task", Root: r.Root, Name: name, Status: now.status.String(),
				DurationMs: now.durationMs, Note: task.Note,
			})
		}

		// A poll is a batch, so the order within one is chosen rather than observed. A task
		// that has started is announced before its output and one that has finished after
		// it, which is the order the two actually happen in — a consumer never sees `Ok`
		// followed by more lines from the same task.
		if changed && !settled(now.status) {
			announce()
			changed = false
		}
		if s.Lines {
			// Where the sent count lands in the buffer as it stands now. Negative means
			// the lines it referred to have been dropped, so start from what is left.
			at := max(0, d.seen[name]-task.Dropped)
			for i := at; i < len(task.Lines); i++ {
				s.Send(Line{
					Type: "line", Root: r.Root, Task: name, Index: task.Dropped + i,
					// A command echo goes out as the command: `command: true` and `task`
					// already carry what the `task: [name] ` prefix was there to say, and
					// a consumer that had to strip it is one that would forget to.
					Text: run.CommandText(task.Lines[i]), Command: task.Lines[i].IsCommand,
				})
			}
			d.seen[name] = task.Dropped + len(task.Lines)
		}
		if changed {
			announce()
		}
	}

	if text, waiting := r.PendingPrompt(); waiting && text != d.prompt {
		d.prompt = text
		s.Send(Prompt{Type: "prompt", Root: r.Root, Text: text})
	}
}

// Finish sends the run's closing event, once. saved is where it was archived, if it was.
func (d *Deltas) Finish(s *Sink, r *run.Run, saved string) {
	if d.closed || s == nil || r == nil {
		return
	}
	d.closed = true
	code := -1
	if r.HasExit {
		code = r.Exit
	}
	s.Send(Exit{
		Type: "exit", Root: r.Root, Code: code,
		DurationMs: r.Duration.Milliseconds(),
		Saved:      saved, Secrets: r.RedactedSecrets,
	})
}

// Done reports whether this run's closing event has gone out.
func (d *Deltas) Done() bool { return d.closed }

// settled reports whether a status is one a task does not come back from.
func settled(s run.Status) bool {
	return s == run.Ok || s == run.Failed || s == run.Skipped
}
